lib/spear/connection/request.ex

```diff
@@ -0,0 +1,194 @@
+defmodule Spear.Connection.Request do
+  @moduledoc false
+
+  # a struct representing a stream-able request
+
+  @type t :: %{
+          continuation: Enumerable.continuation(),
+          request_ref: Mint.Types.request_ref(),
+          buffer: binary(),
+          from: GenServer.from(),
+          response: Spear.Connection.Response.t(),
+          status: :streaming | :done
+        }
+
+  defstruct [:continuation, :buffer, :request_ref, :from, :response, :status]
+
+  def new(event_stream, request_ref, from) do
+    reducer = &reduce_with_suspend/2
+    stream = Stream.map(event_stream, &Spear.Request.to_wire_data/1)
+    continuation = &Enumerable.reduce(stream, &1, reducer)
+
+    %__MODULE__{
+      continuation: continuation,
+      buffer: <<>>,
+      request_ref: request_ref,
+      from: from,
+      response: %Spear.Connection.Response{},
+      status: :streaming
+    }
+  end
+
+  defp reduce_with_suspend(
+         {message, message_size},
+         {message_buffer, message_buffer_size, max_size}
+       )
+       when message_size + message_buffer_size > max_size do
+    {:suspend,
+     {[{message, message_size} | message_buffer], message_size + message_buffer_size, max_size}}
+  end
+
+  defp reduce_with_suspend(
+         {message, message_size},
+         {message_buffer, message_buffer_size, max_size}
+       ) do
+    {:cont,
+     {[{message, message_size} | message_buffer], message_size + message_buffer_size, max_size}}
+  end
+
+  @spec emit_messages(%Spear.Connection{}, %__MODULE__{}) ::
+          {:ok, %Spear.Connection{}} | {:error, %Spear.Connection{}, reason :: any()}
+  def emit_messages(state, %__MODULE__{buffer: <<>>, continuation: continuation} = request) do
```

Comment:
it's not super obvious (especially because it doesn't work like this for `%{}`) but `<<>>` is a match on only the empty binary/string

```elixir
iex> match?(<<>>, "hello")
false
iex> match?(<<>>, "")
true
iex> match?(%{}, %{"fizz" => "buzz"})
true
iex> match?(%{}, %{})
true
```

---

#### Why do we need to buffer stuff at all?

We have the `request.continuation` function which will unfold more events out of the stream, but the way that `reduce_with_suspend/2` is currently written (and must be written unless I'm mistaken) is that we read until we _exceed_ our smallest window

Why wait until we exceed?

1. a single message may itself exceed one or more window(s)
2. there's no way to read-ahead in the stream or put an element we've already reduced out of the stream back into the stream
    - it's not intuitively obvious why this is the case but I can expound if you're interested

so we

1. read the stream until we exceed the window
    - we **suspend** the stream upon exceeding the window
2. write all available bytes up to the window size
    - we can and do write partial messages
    - according to the gRPC spec: "DATA frame boundaries have no relation to **Length-Prefixed-Message** boundaries and implementations should make no assumptions about their alignment." so this is allowed, possible, and correct
3. buffer remaining bytes and the next continuation function in the `%__MODULE__{}` request struct
4. go back to happily waiting as GenServers do for more info messages
5. if we receive an info message off the wire, the connection will give each stream request in state another attempt